### PR TITLE
add warning if  repository was cloned without submodules

### LINKED ISSUE #770: Need a warning if repository was clone without submodules
See as well https://github.com/yaptide/ui/issues/1401

### PR DESCRIPTION
**Pull request by SzymanskiBartlomiej**

### CODE REVIEW

File: yaptide/application.py [hunk 1]
```diff
@@ -50,6 +54,13 @@ def create_app():
 
     return app
 
+def check_submodules():
+    with open('.gitmodules') as f:
+        for line in f:
+            if 'path' in line:
+                path = line.split('=')[-1].strip()
+                if not os.path.exists(path):
+                    return False
```
Comment (grzanka) [inline yaptide/application.py]:
```suggestion
                if not Path(path):
                    return False
```

why not to use pathlib ?

File: yaptide/application.py [hunk 2]
```diff
@@ -50,6 +54,13 @@ def create_app():
 
     return app
 
+def check_submodules():
+    with open('.gitmodules') as f:
+        for line in f:
+            if 'path' in line:
+                path = line.split('=')[-1].strip()
```
Comment (grzanka) [inline yaptide/application.py]:
```suggestion
            if 'path' in line:
                logging.info("Found submodule line in .gitmodules: %s", line)
                path = line.split('=')[-1].strip()
```

add some logging as well

File: yaptide/application.py [hunk 3]
```diff
@@ -50,6 +54,13 @@ def create_app():
 
     return app
 
+def check_submodules():
+    with open('.gitmodules') as f:
+        for line in f:
+            if 'path' in line:
```
Comment (grzanka) [inline yaptide/application.py]: This is bad idea, it will find `path` in all comment lines and other unrelated line.

Comment (grzanka) [reply yaptide/application.py]: You can check also idea suggested by chatgpt. What do you think are the pros and cons of each solution ?

```python
import subprocess

def check_submodules(repo_path):
    # Change the current working directory to the repo path
    subprocess.run(["cd", repo_path], check=True, shell=True)
    
    # Run the git submodule status command
    result = subprocess.run(["git", "submodule", "status"], capture_output=True, text=True, check=True)
    
    # Split the output into lines
    lines = result.stdout.strip().split('\n')
    
    # Check each line for uncloned or not updated submodules
    uncloned_submodules = [line for line in lines if line.startswith('-') or line.startswith('+')]
    
    if uncloned_submodules:
        print("There are submodules that haven't been cloned or are not up to date:")
        for submodule in uncloned_submodules:
            print(submodule)
    else:
        print("All submodules are cloned and up to date.")

# Replace '/path/to/repo' with the actual path to your repository
check_submodules('/path/to/repo')
```

**Comment by grzanka** [comment]
Fix deep source issues

**Comment by grzanka** [comment]
image

File: yaptide/application.py [hunk 4]
```diff
@@ -54,15 +50,19 @@ def create_app():
 
     return app
 
+
 def check_submodules():
+    """Function checking the status of Git submodules"""
     try:
-        result = subprocess.run(["git", "submodule", "status"], capture_output=True, text=True, check=True)
+        result = subprocess.run(["/usr/bin/git", "submodule", "status"], capture_output=True, text=True, check=True)
```
Comment (grzanka) [inline yaptide/application.py]: will this work on Windows ?

**Review by grzanka — CHANGES_REQUESTED** [review]
check if the solution works for Windows as well

**Comment by SzymanskiBartlomiej** [comment]
> check if the solution works for Windows as well

Now it works on both

File: yaptide/application.py [hunk 5]
```diff
@@ -16,6 +17,10 @@ def create_app():
     app = Flask(flask_name)
     logging.info("Creating Flask app %s", flask_name)
 
+    check_submodules()
```
Comment (grzanka) [inline yaptide/application.py]: Why `check_submodules()` happens after creation of Flask app object ? is there any specific reason for that ?

I'd do it as the first thing in the create_app() function.

File: yaptide/application.py [hunk 6]
```diff
@@ -16,6 +17,10 @@ def create_app():
...
+
+    from yaptide.routes.main_routes import initialize_routes
```
Comment (grzanka) [inline yaptide/application.py:17]: Why this import is moved here ?

Comment (SzymanskiBartlomiej) [reply yaptide/application.py:17]: When import is on top (and submodules are not loaded) there is error: "ModuleNotFoundError: No module named 'yaptide.converter'".  Thats why it should be imported after submodules are checked

Comment (grzanka) [reply yaptide/application.py:17]: Good reason, but from reading the code itself its hard to guess that. Add a comment above explaining that `main_routes` are importing (in-directly) the converter module which is cloned as submodule. That means that lack of submodule would result in the ModuleNotFoundError.

File: yaptide/application.py (hunk 5 above)
Comment (SzymanskiBartlomiej) [reply yaptide/application.py]: done

File: yaptide/application.py [hunk 7]
```diff
@@ -51,5 +51,27 @@ def create_app():
     return app
 
 
+def find_git_executable():
+    """Function returns git executable path or None"""
+    git_path = shutil.which('git')
+    return git_path
+
+
+def check_submodules():
```
Comment (grzanka) [inline yaptide/application.py]: I've tried this approach in following way:

1. Cloned repo without submodules (on Windows):

```powershell
PS C:\Users\Legion\workspace\1\yaptide> git clone https://github.com/yaptide/yaptide.git
```

and checkout branch `770-need-a-warning-if-repository-was-clone-without-submodules`

2. Started yaptide with the method described in README.md:

```
PS C:\Users\Legion\workspace\1\yaptide> docker compose up -d --build       
```

I've got an error as expected:

```
[+] Running 3/3
 ✔ Container yaptide_redis       Healthy                                                                           0.0s
 ✔ Container yaptide_postgresql  Healthy                                                                           0.0s
 ✘ Container yaptide_flask       Error                                                                             0.0s
dependency failed to start: container yaptide_flask exited (1)
```

But the error message didn't said much:

```python
2024-03-10 17:09:11 Traceback (most recent call last):
2024-03-10 17:09:11   File "/usr/local/bin/flask", line 8, in <module>
2024-03-10 17:09:11     sys.exit(main())
2024-03-10 17:09:11   File "/usr/local/lib/python3.10/site-packages/flask/cli.py", line 1064, in main
2024-03-10 17:09:11     cli.main()
2024-03-10 17:09:11   File "/usr/local/lib/python3.10/site-packages/click/core.py", line 1078, in main
2024-03-10 17:09:11     rv = self.invoke(ctx)
2024-03-10 17:09:11   File "/usr/local/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
2024-03-10 17:09:11     return _process_result(sub_ctx.command.invoke(sub_ctx))
2024-03-10 17:09:11   File "/usr/local/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
2024-03-10 17:09:11     return ctx.invoke(self.callback, **ctx.params)
2024-03-10 17:09:11   File "/usr/local/lib/python3.10/site-packages/click/core.py", line 783, in invoke
2024-03-10 17:09:11     return __callback(*args, **kwargs)
2024-03-10 17:09:11   File "/usr/local/lib/python3.10/site-packages/click/decorators.py", line 92, in new_func
2024-03-10 17:09:11     return ctx.invoke(f, obj, *args, **kwargs)
2024-03-10 17:09:11   File "/usr/local/lib/python3.10/site-packages/click/core.py", line 783, in invoke
2024-03-10 17:09:11     return __callback(*args, **kwargs)
2024-03-10 17:09:11   File "/usr/local/lib/python3.10/site-packages/flask/cli.py", line 912, in run_command
2024-03-10 17:09:11     raise e from None
2024-03-10 17:09:11   File "/usr/local/lib/python3.10/site-packages/flask/cli.py", line 898, in run_command
2024-03-10 17:09:11     app = info.load_app()
2024-03-10 17:09:11   File "/usr/local/lib/python3.10/site-packages/flask/cli.py", line 309, in load_app
2024-03-10 17:09:11     app = locate_app(import_name, name)
2024-03-10 17:09:11   File "/usr/local/lib/python3.10/site-packages/flask/cli.py", line 236, in locate_app
2024-03-10 17:09:11     return find_best_app(module)
2024-03-10 17:09:11   File "/usr/local/lib/python3.10/site-packages/flask/cli.py", line 64, in find_best_app
2024-03-10 17:09:11     app = app_factory()
2024-03-10 17:09:11   File "/usr/local/app/yaptide/application.py", line 20, in create_app
2024-03-10 17:09:11     check_submodules()
2024-03-10 17:09:11   File "/usr/local/app/yaptide/application.py", line 66, in check_submodules
2024-03-10 17:09:11     raise RuntimeError("Git executable not found.")
2024-03-10 17:09:11 RuntimeError: Git executable not found.
```

It seems the approach with git is OK for developer-mode, but if a user would like to deploy the app, it won't help.

Comment (grzanka) [reply yaptide/application.py]: Obviously there is no `git` binary in docker container used for deployment.

Comment (grzanka) [reply yaptide/application.py]: In comment given here: https://github.com/yaptide/yaptide/pull/875#discussion_r1512805385 I've asked about pros and cons. It seems that is first `cons`. Better to consider `pros` and `cons` before starting implementation :)

Comment (grzanka) [reply yaptide/application.py]: As for another proposition - lets stick to git as it seems quite flexible solution. 
I think that `check_submodule` could be moved into standalone python file, located in say `yaptide/admin` package.
Then you could call it from `Dockerfile-flask` or other to ensure that submodules were cloned.
The logic can stay as it is.

File: yaptide/admin/git_submodules.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    """Function checking if submodules folders exists on system"""
+    with open('.gitmodules') as f:
+        logging.info("Found .gitmodules file")
+        for line in f:
+            if 'path' in line:
+                logging.info("Found submodule line in .gitmodules: %s", line)
+                submodule_path = line.split('=')[-1].strip()
+
+                # Check if folder exists and is not empty
+                if not os.path.exists(submodule_path) or not os.listdir(submodule_path):
```
Comment (grzanka) [inline yaptide/admin/git_submodules.py]: Convert that approach to pathlib, we should use it consistently through all the code

File: yaptide/admin/git_submodules.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+                    raise RuntimeError("Submodules are missing! Please use: git submodule update --init --recursive")
```
Comment (grzanka) [inline yaptide/admin/git_submodules.py]: Mention in the message which submodule is missing.

File: yaptide/application.py [hunk 10]
```diff
@@ -5,11 +5,17 @@
...
 from yaptide.persistence.database import db
-from yaptide.routes.main_routes import initialize_routes
+from yaptide.admin import git_submodules
 
 
 def create_app():
     """Function starting Flask Server"""
+    git_submodules.check_submodules()
+
+    # Main_routes is importing (in-directly) the converter module which is cloned as submodule.
```
Comment (grzanka) [inline yaptide/application.py]:
```suggestion
    # main_routes module is importing (in-directly) the converter module which is cloned as submodule.
```

Comment (SzymanskiBartlomiej) [reply yaptide/application.py]: done

File: yaptide/admin/git_submodules.py (hunk 9 above)
Comment (SzymanskiBartlomiej) [reply yaptide/admin/git_submodules.py]: done

File: yaptide/admin/git_submodules.py (hunk 8 above)
Comment (SzymanskiBartlomiej) [reply yaptide/admin/git_submodules.py]: done

**Review by grzanka — APPROVED** [review]

**Review by grzanka — APPROVED** [review]
